File: Algorithms/133.CloneGraph.java
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode graph) {
        if (graph == null) return null;
        Map<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<>();
        return DFS(graph, map);
    }
    private UndirectedGraphNode DFS(UndirectedGraphNode graph,
            Map<UndirectedGraphNode, UndirectedGraphNode> map) {
        if (map.containsKey(graph)) {
            return map.get(graph);
        }
        UndirectedGraphNode graphCopy = new UndirectedGraphNode(graph.label);
        map.put(graph, graphCopy);
        for (UndirectedGraphNode neighbor : graph.neighbors) {
            graphCopy.neighbors.add(DFS(neighbor, map));
        }
        return graphCopy;
    }
    /*
    // bfs
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) return null;
        Map<UndirectedGraphNode, UndirectedGraphNode>  map = new HashMap<>();
        List<UndirectedGraphNode> q = new LinkedList<>();
        q.add(node);
        UndirectedGraphNode graphCopy = new UndirectedGraphNode(node.label);
        map.put(node, graphCopy);

        while (!q.isEmpty()) {
            node = q.remove();
            for (UndirectedGraphNode neighbor: node.neighbors) {
                if (map.get(neighbor) == null) {
                    UndirectedGraphNode p = new UndirectedGraphNode(neighbor.label);
                    map.put(neighbor, p);
                    q.add(neighbor);
                    map.get(node).neighbors.add(p);
                } else {
                    map.get(node).neighbors.add(map.get(neighbor));
                }
            }
        }
        return graphCopy;
    }
    */
}

File: Algorithms/133.CloneGraph.cpp
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (!node) return NULL;
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> map;
        return dfs(node, map);
    }
    UndirectedGraphNode *dfs(UndirectedGraphNode* node, unordered_map<UndirectedGraphNode *, UndirectedGraphNode *>& map) {
        if (map.find(node) != map.end()) return map[node];
        UndirectedGraphNode* copyNode = new UndirectedGraphNode(node->label);
        map[node] = copyNode;
        for (int i = 0; i < node->neighbors.size(); i++) {
            copyNode->neighbors.push_back(dfs(node->neighbors[i], map));
        }
        return copyNode;
    }
    /*
    // bfs
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (!node) return NULL;
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> map;
        queue<UndirectedGraphNode *> q;
        q.push(node);
        UndirectedGraphNode *graphCopy = new UndirectedGraphNode(node->label);
        map[node] = graphCopy;

        while (!q.empty()) {
            node = q.front();
            q.pop();
            int n = node->neighbors.size();
            for (int i = 0; i < n; i++) {
                UndirectedGraphNode *neighbor = node->neighbors[i];
                if (map.find(neighbor) == map.end()) {
                    UndirectedGraphNode *p = new UndirectedGraphNode(neighbor->label);
                    map[neighbor] = p;
                    q.push(neighbor);
                    map[node]->neighbors.push_back(p);
                } else {
                    map[node]->neighbors.push_back(map[neighbor]);
                }
            }
        }
        return graphCopy;
    }
    */
};

